### PR TITLE
Fixed logic with connection checks

### PR DESCRIPTION
**Pull request by tiago-loureiro**
This is actually an existing bug with the current permission check that surfaced with the latest conversation role PR. The _actual_ logic should be:

```
    -- users can add other users to a conversation iff:
    --    - *the adding user is connected to the users being added*
    --    OR
    --    - *the adding user is part of the team of the users being added*
```

### CODE REVIEW

File: services/galley/test/integration/API/Teams.hs [hunk 1]
```diff
@@ -62,7 +62,7 @@ tests s = testGroup "Teams API"
     , test s "add team conversation as partner (fail)" testAddTeamConvAsExternalPartner
     , test s "add managed conversation through public endpoint (fail)" testAddManagedConv
     , test s "add managed team conversation ignores given users" testAddTeamConvWithUsers
-    , test s "add team member to conversation without connection" testAddTeamMemberToConv
+    , test s "add team member to conversation without connection" (testAddTeamMemberToConv >> ensureQueueEmpty)
```
Comment (fisx) [inline services/galley/test/integration/API/Teams.hs]:
```suggestion
    , test s "add team member to conversation without connection" testAddTeamMemberToConv
```

(this is already happening at the end of the test)

File: services/galley/test/integration/API/Teams.hs [hunk 2]
```diff
@@ -559,31 +559,89 @@ testAddTeamConvWithUsers = do
...
-    Util.assertNotConvMember (mem1^.userId) cid
-    Util.postMembers (mem1^.userId) (list1 (mem2^.userId) []) cid !!! const 404 === statusCode
-    Util.assertNotConvMember (mem2^.userId) cid
+    Util.assertNotConvMember (mem1T1^.userId) cidT1
+    Util.postMembers (mem1T1^.userId) (list1 (mem2T1^.userId) []) cidT1 !!! const 404 === statusCode
+    Util.assertNotConvMember (mem2T1^.userId) cidT1
 
-    -- OTOH, team member 3 _can_ add another team member despite lacking the required team permission
+    -- OTOH, mem3T1 _can_ add another team member despite lacking the required team permission
     -- However, all users are admins by default
```
Comment (fisx) [inline services/galley/test/integration/API/Teams.hs]: i don't understand this line.  should that now read "conversation admins"?

File: services/galley/test/integration/API/Teams.hs [hunk 3]
```diff
@@ -559,31 +559,89 @@ testAddTeamConvWithUsers = do
...
-    Util.assertConvMember owner cid
-    Util.postMembers owner (list1 (mem2^.userId) []) cid !!! const 200 === statusCode
-    Util.assertConvMember (mem2^.userId) cid
+    Util.assertConvMember ownerT1 cidT1
```
Comment (fisx) [inline services/galley/test/integration/API/Teams.hs]:
```suggestion
    Util.assertNotConvMember (mem2T1^.userId) cidT1
```

(or maybe both?)

File: services/galley/test/integration/API/Teams.hs [hunk 4]
```diff
@@ -559,31 +559,89 @@ testAddTeamConvWithUsers = do
...
+    --    OR
+    --    - *the adding user is part of the team of the users being added*
+
+    -- Now we add someone from T2 that we are connected to
+    Util.postMembers ownerT1 (list1 ownerT2 []) cidT1 !!! const 200 === statusCode
+    Util.assertConvMember ownerT2 cidT1
+    -- And they can add their own team members
+    Util.postMembers ownerT2 (list1 (mem1T2^.userId) []) cidT1 !!! const 200 === statusCode
+    Util.assertConvMember (mem1T2^.userId) cidT1
+    -- Still, they cannot add random members from T1, despite the conversation being "hosted" there
```
Comment (fisx) [inline services/galley/test/integration/API/Teams.hs]:
```suggestion
    -- Still, they cannot add random members without a connection from T1, despite the conversation being "hosted" there
```

File: services/galley/test/integration/API/Teams.hs [hunk 5]
```diff
@@ -559,31 +559,89 @@ testAddTeamConvWithUsers = do
...
+    Util.postMembers ownerT2 (list1 (mem4T1^.userId) []) cidT1 !!! const 403 === statusCode
+    Util.assertNotConvMember (mem4T1^.userId) cidT1
+
+    -- Now let's look at convs hosted on team2
+    -- ownerT2 *is* connected to ownerT1
+    Util.postMembers ownerT2 (list1 ownerT1 []) cidT2 !!! const 200 === statusCode
+    Util.assertConvMember ownerT2 cidT2
```
Comment (fisx) [inline services/galley/test/integration/API/Teams.hs]:
```suggestion
    Util.assertConvMember ownerT1 cidT2
```

File: services/galley/test/integration/API/Teams.hs [hunk 6]
```diff
@@ -559,31 +559,89 @@ testAddTeamConvWithUsers = do
...
+    -- and mem1T2 is on the same team, but mem1T1 is *not*
+    Util.postMembers ownerT2 (list1 (mem1T2^.userId) [mem1T1^.userId]) cidT2 !!! const 403 === statusCode
```
Comment (fisx) [inline services/galley/test/integration/API/Teams.hs]: after this line:

```haskell
     Util.assertNotConvMember (mem1T1^.userId) cidT2
     Util.assertNotConvMember (mem1T2^.userId) cidT2
```

File: services/galley/test/integration/API/Teams.hs [hunk 7]
```diff
@@ -559,31 +559,89 @@ testAddTeamConvWithUsers = do
...
+    -- mem1T2 is on the same team, so that is fine too
+    Util.postMembers ownerT2 (list1 (mem1T2^.userId) []) cidT2 !!! const 200 === statusCode
+    Util.assertConvMember (mem1T2^.userId) cidT2
+    -- ownerT2 is *NOT* connected to mem3T1 and not on the same team, so should not be allowed to add
+    Util.postMembers ownerT2 (list1 (mem3T1^.userId) []) cidT2 !!! const 403 === statusCode
+    Util.assertNotConvMember (mem3T1^.userId) cidT2
+
+    -- For personal conversations, same logic applies
+
+    -- Can add users that am connected to
```
Comment (fisx) [inline services/galley/test/integration/API/Teams.hs]:
```suggestion
    -- Can add connected users
```

File: services/galley/src/Galley/API/Util.hs [hunk 8]
```diff
@@ -43,6 +43,21 @@ ensureAccessRole role users mbTms = case role of
...
+-- for non-team-members of the _given_ user
+ensureConnectedOrSameTeam :: UserId -> [UserId] -> Galley ()
+ensureConnectedOrSameTeam _ []   = pure ()
+ensureConnectedOrSameTeam u uids = do
+    uTeams <- Data.userTeams u
+    -- We collect all the relevant uids from same teams as the origin user
+    sameTeamUids <- forM uTeams $ \team ->
+        fmap (view userId) <$> Data.teamMembersLimited team uids
+    -- Do not check connections for users that are on the same team
+    ensureConnected u (uids \\ (join sameTeamUids))
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Util.hs]: 👍

File: services/galley/src/Galley/API/Util.hs [hunk 9]
```diff
@@ -43,6 +43,21 @@ ensureAccessRole role users mbTms = case role of
         when (length activated /= length users) $ throwM convAccessDenied
     NonActivatedAccessRole -> return ()
 
+-- | Check that the given user is either part of the same team(s) as the other
+-- users OR that there is a connection.
+--
+-- Team members are always considered connected, so we only check 'ensureConnected'
+-- for non-team-members of the _given_ user
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Util.hs]: I checked all the other cases, the others are correct; they related to creation of specific conversations while this case relates to adding to _any type_ of conversation

File: services/galley/test/integration/API/Teams.hs (hunk 1 above)
Comment (tiago-loureiro) [reply services/galley/test/integration/API/Teams.hs]: > (this is already happening at the end of the test)

Is it? Where? 😬

File: services/galley/test/integration/API/Teams.hs (hunk 2 above)
Comment (tiago-loureiro) [reply services/galley/test/integration/API/Teams.hs]: Right, let me rephrase that

File: services/galley/test/integration/API/Teams.hs (hunk 3 above)
Comment (tiago-loureiro) [reply services/galley/test/integration/API/Teams.hs]: Ah no; if u look better, we're just ensuring `ownerT1` is a member of the conv before adding other members

File: services/galley/test/integration/API/Teams.hs (hunk 6 above)
Comment (tiago-loureiro) [reply services/galley/test/integration/API/Teams.hs]: 👍

**Comment by tiago-loureiro** [comment]
All tests passed for me locally; I will merge this as-is and fix the CI after.

File: services/galley/test/integration/API/Teams.hs (hunk 1 above)
Comment (fisx) [reply services/galley/test/integration/API/Teams.hs]: oops, i may have skipped to [this line](https://github.com/wireapp/wire-server/blob/e4b5e1d9050ca3496ffa4d8f0d35482fb4aa97c5/services/galley/test/integration/API/Teams.hs#L711), which is a different test case.

anyway, it should be in the test IMO.  :)

(not a big deal)
